pageserver/src/pgdatadir_mapping.rs

```diff
@@ -1350,28 +1356,67 @@ impl<'a> DatadirModification<'a> {
         content: &[u8],
         ctx: &RequestContext,
     ) -> anyhow::Result<()> {
-        let mut dir = match self.get(AUX_FILES_KEY, ctx).await {
-            Ok(buf) => AuxFilesDirectory::des(&buf)?,
-            Err(e) => {
-                // This is expected: historical databases do not have the key.
-                debug!("Failed to get info about AUX files: {}", e);
-                AuxFilesDirectory {
-                    files: HashMap::new(),
-                }
-            }
+        let key = path.to_string();
+        let value = if content.is_empty() {
+            None
+        } else {
+            Some(Bytes::copy_from_slice(content))
         };
-        let path = path.to_string();
-        if content.is_empty() {
-            dir.files.remove(&path);
+
+        if self.pending_aux_files {
+            // We already updated aux files in `self`: emit a delta and update our latest value
+
+            self.put(
+                AUX_FILES_KEY,
+                Value::WalRecord(NeonWalRecord::AuxFile {
+                    key: key.clone(),
+                    value: value.clone(),
+                }),
+            );
         } else {
-            dir.files.insert(path, Bytes::copy_from_slice(content));
+            // Check if the AUX_FILES_KEY is initialized
+            match self.get(AUX_FILES_KEY, ctx).await {
+                Ok(_) => {
+                    // Key is already set, we may append a delta
+                    self.put(
+                        AUX_FILES_KEY,
+                        Value::WalRecord(NeonWalRecord::AuxFile {
+                            key: key.clone(),
+                            value: value.clone(),
+                        }),
+                    );
+                }
+                Err(
+                    e @ (PageReconstructError::AncestorStopping(_)
+                    | PageReconstructError::Cancelled
+                    | PageReconstructError::AncestorLsnTimeout(_)),
+                ) => {
+                    // Important that we do not interpret a shutdown error as "not found" and thereby
+                    // reset the map.
+                    return Err(e.into());
+                }
+                // FIXME: PageReconstructError doesn't have an explicit variant for key-not-found, so
```

Comment:
This FIXME is calling out an existing bug that I noticed while modifying the code.